### PR TITLE
Bugfix - compute scale_inv when loading checkpoint

### PR DESCRIPTION
**Pull request by galagam**
On torch model loading, update fp8 scale_inv from the extra state tensor. 
+ test to validate all extra state attributes are consistent after torch load.

### CODE REVIEW

**Comment by galagam** [comment]
@nzmora-nvidia for visibility

**Comment by ksivaman** [comment]
Scaling factor inverses are updated first thing in the following iteration which is why we don't currently do this. Are you observing different behavior?

**Comment by galagam** [comment]
@ksivaman 
I'm continuing Asfiya's work on enabling GPT-3 export to ONNX.
The bug observed is that Q/DQ scale is always 1 in this flow:
1. Train & save PT checkpoint
2. Load PT checkpoint
3. Export to ONNX
This happens because the onnx_cast_to_fp8() and onnx_cast_from_fp8() use the scale_inv parameter, but when loading a checkpoint we're initializing it to ones, and then not setting it according to the value saved in extra_state.

This can be observed under 
https://gitlab-master.nvidia.com/ghubaraagam/megatron-lm/-/commits/dev-gpt3-export entry point scripts/export_onnx.sh

File: tests/pytorch/test_torch_save_load.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    model_out = Test_TE_Export(precision, True)
+    model_out.load_state_dict(torch.load(tmp_filename))
+    model_out.eval()
+
+    assert torch.allclose(model_in.fp8_meta["scaling_fwd"].scale, model_out.fp8_meta["scaling_fwd"].scale)
+    assert torch.allclose(model_in.fp8_meta["scaling_fwd"].scale_inv, model_out.fp8_meta["scaling_fwd"].scale_inv)
+    assert torch.allclose(model_in.fp8_meta["scaling_fwd"].amax_history, model_out.fp8_meta["scaling_fwd"].amax_history)
+    assert torch.allclose(model_in.fp8_meta["scaling_bwd"].scale, model_out.fp8_meta["scaling_bwd"].scale)
+    assert torch.allclose(model_in.fp8_meta["scaling_bwd"].scale_inv, model_out.fp8_meta["scaling_bwd"].scale_inv)
+    assert torch.allclose(model_in.fp8_meta["scaling_bwd"].amax_history, model_out.fp8_meta["scaling_bwd"].amax_history)
```
Comment (nzmora-nvidia) [inline tests/pytorch/test_torch_save_load.py:116]: These lines have the same content as the 3 lines above

File: tests/pytorch/test_torch_save_load.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+    class Test_TE_Export(TransformerEngineBaseModule):
+        def __init__(self, precision, use_bias):
+            super().__init__()
+            self.use_bias = use_bias
+            self.precision = precision
+
+            self.fp8_tensor_inp = tex.FP8FwdTensors.GEMM1_INPUT
+            self.fp8_tensor_weight = tex.FP8FwdTensors.GEMM1_WEIGHT
+            nb_inp_scales, nb_weight_scales = 1, out_features
```
Comment (nzmora-nvidia) [inline tests/pytorch/test_torch_save_load.py:55]: TE does per-tensor scaling so `nb_weight_scales` should also be set to 1

**Review by nzmora-nvidia — COMMENTED** [review]
LGTM

File: tests/pytorch/test_torch_save_load.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+    model_in = Test_TE_Export(precision, True)
+    with te.fp8_autocast(enabled=True):
+        model_in.fp8_init()
+        model_in.fp8_meta["scaling_fwd"].scale = torch.ones(3, dtype=torch.float32, device="cuda") * scale_fwd
+        model_in.fp8_meta["scaling_fwd"].scale_inv = torch.ones(3, dtype=torch.float32, device="cuda") / scale_fwd
+        model_in.fp8_meta["scaling_fwd"].amax_history = torch.ones(3, dtype=torch.float32, device="cuda") * history_fwd
+        model_in.fp8_meta["scaling_bwd"].scale = torch.ones(2, dtype=torch.float32, device="cuda") * scale_bwd
+        model_in.fp8_meta["scaling_bwd"].scale_inv = torch.ones(2, dtype=torch.float32, device="cuda") / scale_bwd
+        model_in.fp8_meta["scaling_bwd"].amax_history = torch.ones(2, dtype=torch.float32, device="cuda") * history_bwd
```
[the remark this answers is not in the document]
Comment (galagam) [reply tests/pytorch/test_torch_save_load.py:103]: It's scaling BWD and scaling FWD. 
I'm only checking the save/load mechanism, so there's no dependency between these values for the purpose of this test. I can refactor to make it clearer.

[the remark this answers is not in the document]
Comment (nzmora-nvidia) [reply tests/pytorch/test_torch_save_load.py:103]: Thanks - no need to refactor, but maybe put a small remark.
I had a feeling I was missing something. I read these lines at least 10 times looking for some diff, but my brain didn't pick  it up.

**Review by nzmora-nvidia — APPROVED** [review]

File: tests/pytorch/test_torch_save_load.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (galagam) [reply tests/pytorch/test_torch_save_load.py:104]: added comment to clarify

File: tests/pytorch/test_torch_save_load.py (hunk 1 above)
Comment (galagam) [reply tests/pytorch/test_torch_save_load.py:119]: added comment to clarify

**Comment by galagam** [comment]
> Scaling factor inverses are updated first thing in the following iteration which is why we don't currently do this. Are you observing different behavior?

As discussed, save and load checkpoint should retain the same state, without requiring additional steps in between. 
In that spirit, I modified get_extra_state() i.o.t save inverse scale separately, and not compute it during set_extra_state(), to avoid numeric differences on different machines and versions. 
In case of older checkpoints where inverse scale is not saved, it will be computed from the scale.

File: transformer_engine/pytorch/module.py [hunk 4]
```diff
@@ -292,8 +294,10 @@ def set_extra_state(self, state: torch.Tensor) -> None:
             # Initialize before loading
             self.init_fp8_meta_tensors()
             self.fp8_meta["scaling_fwd"].scale.copy_(scale_fwd)
+            self.fp8_meta["scaling_fwd"].scale_inv.copy_(1.0/scale_fwd)
```
Comment (ksivaman) [inline transformer_engine/pytorch/module.py:297]: This is the backwards compatibility path for a previous version and should not be changed. Please remove.

File: transformer_engine/pytorch/module.py [hunk 5]
```diff
@@ -292,8 +294,10 @@ def set_extra_state(self, state: torch.Tensor) -> None:
...
             self.fp8_meta["scaling_fwd"].amax_history.copy_(amax_history_fwd)
             self.fp8_meta["scaling_bwd"].scale.copy_(scale_bwd)
+            self.fp8_meta["scaling_bwd"].scale_inv.copy_(1.0/scale_bwd)
```
Comment (ksivaman) [inline transformer_engine/pytorch/module.py:300]: Same as above

File: transformer_engine/pytorch/module.py (hunk 4 above)
Comment (galagam) [reply transformer_engine/pytorch/module.py]: Fair enough. Fixed (amended commit 9dac81c).

File: transformer_engine/pytorch/module.py (hunk 5 above)
Comment (galagam) [reply transformer_engine/pytorch/module.py]: Fair enough. Fixed (amended commit 9dac81c).

**Review by ksivaman — APPROVED** [review]
LGTM

**Pull request merged**